### ISSUE TITLE
Typed Boxes - extend TypeAdapter to Boxes

### ISSUE THREAD

**Issue by daniel-v** [hivedb/hive#75]
We can handle values stored in a typed manner now via TypeAdapter;, in my head it makes sense to extend the TypeAdapter concept to Boxes.

## Use-cases

I know beforehand that I'll want to store only a single type in a box. 

## Proposal

Make read/write operations accept typed value.

eg:

```dart
TypedModel model = TypedModel();
box.add(model);
box.add('string'); // should throw error that string is not a subtype of TypedModel
// or
TypedModel readValue = box.get('key'); // should be valid
```

I created a small proof of concept that you guys can try out and experiment with this a little bit:
https://gist.github.com/daniel-v/069aa68fe0ae4f2d1c25bc16bee560ca

From this implementation, there is 1 challenge to solve: how to store the type id for types.

**Comment by leisim** [comment]
I agree, it would be very nice to have some kind of "TypedBox". The prototype of Hive even had something similar. The problem was that I did not manage to make it work with lists:

You cannot do the following in dart: `List<dynamic> as List<String>`. This is not only true for lists but all generic objects.

If we would say that you just cannot use a typed box with generic types, this feature would be very easy to implement. I don't even think a `TypedBox` type would be needed. Adding a type parameter `Box<T>` to the box interface is not a breaking change.

**Comment by daniel-v** [comment]
No, you indeed can't. 

Though I'm wondering, where exactly do you need to rely on that kind of cast?

For Lists and Maps, there is `List.cast`/`Map.cast` or `List.of()`/`Map.of()`. These are overhead but should guarantee runtime type safety.

Is the argument that such a type would fail?

```dart
class BoxType<T> {
  List<T> someprops;
}
```

**Comment by Afsar-Pasha** [comment]
@daniel-v see #45

**Comment by daniel-v** [comment]
Didn't disappear, looking for solutions whenever I got time.

**Comment by leisim** [comment]
This has been resolved in `v1.1.0`. You can now do the following: `Hive.openBox<User>()` and you will get a `Box<User>`.
